File: siden-core/src/main/java/ninja/siden/internal/LambdaWebSocketFactory.java
/*
 * Copyright 2014 SATO taichi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package ninja.siden.internal;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;

import ninja.siden.Connection;
import ninja.siden.WebSocket;
import ninja.siden.WebSocketCustomizer;
import ninja.siden.WebSocketFactory;
import ninja.siden.util.ExceptionalBiConsumer;
import ninja.siden.util.ExceptionalConsumer;

/**
 * @author taichi
 */
public class LambdaWebSocketFactory implements WebSocketFactory,
		WebSocketCustomizer {

	List<ExceptionalConsumer<Connection, Exception>> conn = new ArrayList<>();
	List<ExceptionalBiConsumer<Connection, String, Exception>> txt = new ArrayList<>();
	List<ExceptionalBiConsumer<Connection, ByteBuffer[], Exception>> bin = new ArrayList<>();
	List<ExceptionalBiConsumer<Connection, ByteBuffer[], Exception>> pong = new ArrayList<>();
	List<ExceptionalBiConsumer<Connection, ByteBuffer[], Exception>> ping = new ArrayList<>();
	List<ExceptionalBiConsumer<Connection, ByteBuffer[], Exception>> close = new ArrayList<>();

	@Override
	public WebSocket create(Connection connection) {
		return new LambdaWebSocket();
	}

	class LambdaWebSocket implements WebSocket {
		Connection connection;

		List<ExceptionalConsumer<Connection, Exception>> conn = new ArrayList<>(
				LambdaWebSocketFactory.this.conn);
		List<ExceptionalBiConsumer<Connection, String, Exception>> txt = new ArrayList<>(
				LambdaWebSocketFactory.this.txt);
		List<ExceptionalBiConsumer<Connection, ByteBuffer[], Exception>> bin = new ArrayList<>(
				LambdaWebSocketFactory.this.bin);
		List<ExceptionalBiConsumer<Connection, ByteBuffer[], Exception>> pong = new ArrayList<>(
				LambdaWebSocketFactory.this.pong);
		List<ExceptionalBiConsumer<Connection, ByteBuffer[], Exception>> ping = new ArrayList<>(
				LambdaWebSocketFactory.this.ping);
		List<ExceptionalBiConsumer<Connection, ByteBuffer[], Exception>> close = new ArrayList<>(
				LambdaWebSocketFactory.this.close);

		@Override
		public void onConnect(Connection connection) throws Exception {
			this.connection = connection;
			for (ExceptionalConsumer<Connection, Exception> fn : this.conn) {
				fn.accept(this.connection);
			}
		}

		<T> void forEach(T payload,
				List<ExceptionalBiConsumer<Connection, T, Exception>> list)
				throws Exception {
			for (ExceptionalBiConsumer<Connection, T, Exception> fn : list) {
				fn.accept(this.connection, payload);
			}
		}

		@Override
		public void onText(String payload) throws Exception {
			forEach(payload, this.txt);
		}

		@Override
		public void onBinary(ByteBuffer[] payload) throws Exception {
			forEach(payload, this.bin);
		}

		@Override
		public void onPong(ByteBuffer[] payload) throws Exception {
			forEach(payload, this.pong);
		}

		@Override
		public void onPing(ByteBuffer[] payload) throws Exception {
			forEach(payload, this.ping);
		}

		@Override
		public void onClose(ByteBuffer[] payload) throws Exception {
			forEach(payload, this.close);
		}
	}

	@Override
	public WebSocketCustomizer onConnect(
			ExceptionalConsumer<Connection, Exception> fn) {
		this.conn.add(fn);
		return this;
	}

	@Override
	public WebSocketCustomizer onText(
			ExceptionalBiConsumer<Connection, String, Exception> fn) {
		this.txt.add(fn);
		return this;
	}

	@Override
	public WebSocketCustomizer onBinary(
			ExceptionalBiConsumer<Connection, ByteBuffer[], Exception> fn) {
		this.bin.add(fn);
		return this;
	}

	@Override
	public WebSocketCustomizer onPong(
			ExceptionalBiConsumer<Connection, ByteBuffer[], Exception> fn) {
		this.pong.add(fn);
		return this;
	}

	@Override
	public WebSocketCustomizer onPing(
			ExceptionalBiConsumer<Connection, ByteBuffer[], Exception> fn) {
		this.ping.add(fn);
		return this;
	}

	@Override
	public WebSocketCustomizer onClose(
			ExceptionalBiConsumer<Connection, ByteBuffer[], Exception> fn) {
		this.close.add(fn);
		return this;
	}
}

File: siden-core/src/main/kotlin/ninja/siden/internal/LambdaWebSocketFactory.kt
/*
 * Copyright 2014 SATO taichi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package ninja.siden.internal

import ninja.siden.Connection
import ninja.siden.WebSocket
import ninja.siden.WebSocketCustomizer
import ninja.siden.WebSocketFactory
import java.nio.ByteBuffer
import java.util.*

/**
 * @author taichi
 */
class LambdaWebSocketFactory : WebSocketFactory, WebSocketCustomizer {

    internal fun <T> makeList(): MutableList<(Connection, T) -> Unit> = ArrayList()

    internal var connFns = ArrayList<(Connection) -> Unit>()
    internal var txtFns = makeList<String>()
    internal var binFns = makeList<Array<ByteBuffer>>()
    internal var pongFns = makeList<Array<ByteBuffer>>()
    internal var pingFns = makeList<Array<ByteBuffer>>()
    internal var closeFns = makeList<Array<ByteBuffer>>()

    override fun create(connection: Connection): WebSocket {
        return LambdaWebSocket()
    }

    internal inner class LambdaWebSocket : WebSocket {
        var connection: Connection? = null

        internal fun <T> makeList(src: List<(Connection, T) -> Unit>): MutableList<(Connection, T) -> Unit> = ArrayList(src)

        val conn = ArrayList<(Connection) -> Unit>(this@LambdaWebSocketFactory.connFns)
        val txt = makeList(this@LambdaWebSocketFactory.txtFns)
        val bin = makeList(this@LambdaWebSocketFactory.binFns)
        val pong = makeList(this@LambdaWebSocketFactory.pongFns)
        val ping = makeList(this@LambdaWebSocketFactory.pingFns)
        val close = makeList(this@LambdaWebSocketFactory.closeFns)

        override fun onConnect(connection: Connection) {
            this.connection = connection
            this.conn.forEach { it(connection) }
        }

        fun <T> forEach(payload: T, list: List<(Connection, T) -> Unit>) {
            requireNotNull(this.connection)
            val c = this.connection
            if (c != null) {
                list.forEach { it(c, payload) }
            }
        }

        override fun onText(payload: String) = forEach(payload, this.txt)

        override fun onBinary(payload: Array<ByteBuffer>) = forEach(payload, this.bin)

        override fun onPong(payload: Array<ByteBuffer>) = forEach(payload, this.pong)

        override fun onPing(payload: Array<ByteBuffer>) = forEach(payload, this.ping)

        override fun onClose(payload: Array<ByteBuffer>) = forEach(payload, this.close)
    }

    override fun onConnect(fn: (Connection) -> Unit): WebSocketCustomizer {
        this.connFns.add(fn)
        return this
    }

    override fun onText(fn: (Connection, String) -> Unit): WebSocketCustomizer {
        this.txtFns.add(fn)
        return this
    }

    override fun onBinary(fn: (Connection, Array<ByteBuffer>) -> Unit): WebSocketCustomizer {
        this.binFns.add(fn)
        return this
    }

    override fun onPong(fn: (Connection, Array<ByteBuffer>) -> Unit): WebSocketCustomizer {
        this.pongFns.add(fn)
        return this
    }

    override fun onPing(fn: (Connection, Array<ByteBuffer>) -> Unit): WebSocketCustomizer {
        this.pingFns.add(fn)
        return this
    }

    override fun onClose(fn: (Connection, Array<ByteBuffer>) -> Unit): WebSocketCustomizer {
        this.closeFns.add(fn)
        return this
    }
}
